File: deploy/trinity-nexus/lang/libs/python/trinity_vsa/src/trinity_vsa/ops.py
"""VSA operations: bind, bundle, permute, similarity."""

from typing import List, Sequence
import numpy as np

from .core import TritVector


def bind(a: TritVector, b: TritVector) -> TritVector:
    """Bind two vectors (element-wise multiplication).
    
    Creates an association between two concepts.
    
    Properties:
        - bind(a, a) results in all +1 (for non-zero elements)
        - bind(a, bind(a, b)) = b (unbinding)
        - Commutative: bind(a, b) = bind(b, a)
    
    Args:
        a: First vector
        b: Second vector
        
    Returns:
        Bound vector
        
    Example:
        >>> apple = TritVector.random(1000)
        >>> red = TritVector.random(1000)
        >>> red_apple = bind(apple, red)
    """
    assert a.dim == b.dim, "Vectors must have same dimension"
    return TritVector(a.data * b.data)


def unbind(bound: TritVector, key: TritVector) -> TritVector:
    """Unbind (same as bind for balanced ternary).
    
    Retrieves one vector from a bound pair.
    
    Args:
        bound: Bound vector
        key: Key vector to unbind with
        
    Returns:
        Retrieved vector (approximately equal to the other bound vector)
        
    Example:
        >>> red_apple = bind(apple, red)
        >>> recovered = unbind(red_apple, red)
        >>> # recovered ≈ apple
    """
    return bind(bound, key)


def bundle(vectors: Sequence[TritVector]) -> TritVector:
    """Bundle multiple vectors (majority voting).
    
    Creates a superposition that is similar to all inputs.
    
    Args:
        vectors: Sequence of vectors to bundle
        
    Returns:
        Bundled vector
        
    Example:
        >>> fruits = bundle([apple, orange, banana])
        >>> # fruits is similar to all three
    """
    if not vectors:
        return TritVector.zeros(0)
    
    dim = vectors[0].dim
    assert all(v.dim == dim for v in vectors), "All vectors must have same dimension"
    
    # Sum all vectors
    sums = np.sum([v.data.astype(np.int32) for v in vectors], axis=0)
    
    # Threshold: positive -> 1, negative -> -1, zero -> 0
    result = np.sign(sums).astype(np.int8)
    return TritVector(result)


def permute(v: TritVector, shift: int) -> TritVector:
    """Permute vector (circular shift).
    
    Used for encoding sequences and positions.
    
    Args:
        v: Vector to permute
        shift: Number of positions to shift (positive = right)
        
    Returns:
        Permuted vector
        
    Example:
        >>> # Encode sequence: word1, word2, word3
        >>> seq = bind(word1, bind(permute(word2, 1), permute(word3, 2)))
    """
    return TritVector(np.roll(v.data, shift))


def similarity(a: TritVector, b: TritVector) -> float:
    """Cosine similarity between two vectors.
    
    Args:
        a: First vector
        b: Second vector
        
    Returns:
        Similarity in [-1.0, 1.0]:
        - 1.0: identical
        - 0.0: orthogonal
        - -1.0: opposite
        
    Example:
        >>> sim = similarity(red_apple, apple)
        >>> print(f"Similarity: {sim:.3f}")
    """
    assert a.dim == b.dim, "Vectors must have same dimension"
    
    dot = np.sum(a.data.astype(np.float64) * b.data.astype(np.float64))
    norm_a = np.sqrt(np.sum(a.data.astype(np.float64) ** 2))
    norm_b = np.sqrt(np.sum(b.data.astype(np.float64) ** 2))
    
    if norm_a == 0 or norm_b == 0:
        return 0.0
    
    return float(dot / (norm_a * norm_b))


def hamming_distance(a: TritVector, b: TritVector) -> int:
    """Hamming distance (number of differing positions).
    
    Args:
        a: First vector
        b: Second vector
        
    Returns:
        Number of positions where vectors differ
    """
    assert a.dim == b.dim, "Vectors must have same dimension"
    return int(np.sum(a.data != b.data))


def dot_product(a: TritVector, b: TritVector) -> int:
    """Dot product of two vectors.
    
    Args:
        a: First vector
        b: Second vector
        
    Returns:
        Sum of element-wise products
    """
    assert a.dim == b.dim, "Vectors must have same dimension"
    return int(np.sum(a.data.astype(np.int64) * b.data.astype(np.int64)))


def encode_sequence(words: List[TritVector]) -> TritVector:
    """Encode a sequence of vectors using permutation.
    
    Args:
        words: List of word vectors in order
        
    Returns:
        Single vector encoding the sequence
        
    Example:
        >>> sentence = encode_sequence([the, cat, sat, on, mat])
    """
    if not words:
        return TritVector.zeros(0)
    
    result = words[0]
    for i, word in enumerate(words[1:], 1):
        result = bind(result, permute(word, i))
    
    return result


def cleanup(query: TritVector, memory: List[TritVector], threshold: float = 0.0) -> TritVector:
    """Clean up noisy vector by finding closest match in memory.
    
    Args:
        query: Noisy query vector
        memory: List of clean prototype vectors
        threshold: Minimum similarity to return a match
        
    Returns:
        Closest matching vector from memory, or query if no match
    """
    if not memory:
        return query
    
    best_sim = threshold
    best_match = query
    
    for proto in memory:
        sim = similarity(query, proto)
        if sim > best_sim:
            best_sim = sim
            best_match = proto
    
    return best_match

File: src/libs/rust/trinity-vsa/src/ops.rs
//! VSA operations: bind, bundle, permute, similarity

use crate::TritVector;

/// Bind two vectors (element-wise multiplication)
/// 
/// Creates an association between two concepts.
/// 
/// # Properties
/// - `bind(a, a) = all +1` (self-binding)
/// - `bind(a, bind(a, b)) = b` (unbinding)
/// - Commutative: `bind(a, b) = bind(b, a)`
/// 
/// # Example
/// ```
/// use trinity_vsa::{TritVector, bind};
/// 
/// let a = TritVector::random(100);
/// let b = TritVector::random(100);
/// let bound = bind(&a, &b);
/// ```
pub fn bind(a: &TritVector, b: &TritVector) -> TritVector {
    assert_eq!(a.len(), b.len(), "Vectors must have same dimension");
    
    let data: Vec<i8> = a.as_slice()
        .iter()
        .zip(b.as_slice().iter())
        .map(|(&x, &y)| x * y)
        .collect();
    
    TritVector::from_slice(&data)
}

/// Unbind (same as bind for balanced ternary)
/// 
/// Retrieves one vector from a bound pair.
/// 
/// # Example
/// ```
/// use trinity_vsa::{TritVector, bind, unbind, similarity};
/// 
/// let a = TritVector::random(100);
/// let b = TritVector::random(100);
/// let bound = bind(&a, &b);
/// let recovered = unbind(&bound, &a);
/// // recovered ≈ b
/// ```
pub fn unbind(bound: &TritVector, key: &TritVector) -> TritVector {
    bind(bound, key)
}

/// Bundle multiple vectors (majority voting)
/// 
/// Creates a superposition that is similar to all inputs.
/// 
/// # Example
/// ```
/// use trinity_vsa::{TritVector, bundle};
/// 
/// let vectors: Vec<TritVector> = (0..5)
///     .map(|_| TritVector::random(100))
///     .collect();
/// let refs: Vec<&TritVector> = vectors.iter().collect();
/// let bundled = bundle(&refs);
/// ```
pub fn bundle(vectors: &[&TritVector]) -> TritVector {
    if vectors.is_empty() {
        return TritVector::zeros(0);
    }
    
    let dim = vectors[0].len();
    assert!(vectors.iter().all(|v| v.len() == dim), "All vectors must have same dimension");
    
    let mut sums: Vec<i32> = vec![0; dim];
    
    for v in vectors {
        for (i, &trit) in v.as_slice().iter().enumerate() {
            sums[i] += trit as i32;
        }
    }
    
    // Threshold: positive -> 1, negative -> -1, zero -> 0
    let data: Vec<i8> = sums.iter().map(|&s| s.signum() as i8).collect();
    TritVector::from_slice(&data)
}

/// Permute vector (circular shift)
/// 
/// Used for encoding sequences and positions.
/// 
/// # Example
/// ```
/// use trinity_vsa::{TritVector, permute};
/// 
/// let v = TritVector::random(100);
/// let shifted = permute(&v, 5);  // Shift right by 5
/// ```
pub fn permute(v: &TritVector, shift: i32) -> TritVector {
    let len = v.len();
    if len == 0 {
        return TritVector::zeros(0);
    }
    
    let shift = ((shift % len as i32) + len as i32) as usize % len;
    
    let mut data = vec![0i8; len];
    for i in 0..len {
        let new_idx = (i + shift) % len;
        data[new_idx] = v.as_slice()[i];
    }
    
    TritVector::from_slice(&data)
}

/// Cosine similarity between two vectors
/// 
/// Returns value in [-1.0, 1.0]:
/// - 1.0: identical
/// - 0.0: orthogonal
/// - -1.0: opposite
/// 
/// # Example
/// ```
/// use trinity_vsa::{TritVector, similarity};
/// 
/// let a = TritVector::random(100);
/// let b = TritVector::random(100);
/// let sim = similarity(&a, &b);
/// ```
pub fn similarity(a: &TritVector, b: &TritVector) -> f64 {
    assert_eq!(a.len(), b.len(), "Vectors must have same dimension");
    
    let dot: i64 = a.as_slice()
        .iter()
        .zip(b.as_slice().iter())
        .map(|(&x, &y)| (x as i64) * (y as i64))
        .sum();
    
    let norm_a: f64 = a.as_slice().iter().map(|&x| (x as f64).powi(2)).sum::<f64>().sqrt();
    let norm_b: f64 = b.as_slice().iter().map(|&x| (x as f64).powi(2)).sum::<f64>().sqrt();
    
    if norm_a == 0.0 || norm_b == 0.0 {
        return 0.0;
    }
    
    (dot as f64) / (norm_a * norm_b)
}

/// Hamming distance (number of differing positions)
/// 
/// # Example
/// ```
/// use trinity_vsa::{TritVector, hamming_distance};
/// 
/// let a = TritVector::random(100);
/// let b = TritVector::random(100);
/// let dist = hamming_distance(&a, &b);
/// ```
pub fn hamming_distance(a: &TritVector, b: &TritVector) -> usize {
    assert_eq!(a.len(), b.len(), "Vectors must have same dimension");
    
    a.as_slice()
        .iter()
        .zip(b.as_slice().iter())
        .filter(|(&x, &y)| x != y)
        .count()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bind_unbind() {
        let a = TritVector::random(100);
        let b = TritVector::random(100);
        
        let bound = bind(&a, &b);
        let recovered = unbind(&bound, &a);
        
        // recovered should be similar to b
        let sim = similarity(&recovered, &b);
        assert!(sim > 0.9, "Unbind should recover original: {}", sim);
    }

    #[test]
    fn test_bundle_similarity() {
        let vectors: Vec<TritVector> = (0..5)
            .map(|_| TritVector::random(1000))
            .collect();
        let refs: Vec<&TritVector> = vectors.iter().collect();
        
        let bundled = bundle(&refs);
        
        // Bundled should be similar to all inputs
        for v in &vectors {
            let sim = similarity(&bundled, v);
            assert!(sim > 0.3, "Bundle should be similar to inputs: {}", sim);
        }
    }

    #[test]
    fn test_permute() {
        let v = TritVector::from_slice(&[1, -1, 0, 1, -1]);
        let shifted = permute(&v, 2);
        assert_eq!(shifted.as_slice(), &[-1, 1, 1, -1, 0]);
    }

    #[test]
    fn test_self_similarity() {
        let v = TritVector::random(100);
        let sim = similarity(&v, &v);
        assert!((sim - 1.0).abs() < 1e-10, "Self-similarity should be 1.0");
    }
}
